### ISSUE TITLE
Rolling releases

### ISSUE THREAD

**Issue by garygreen** [pstadler/flightplan#113]
I'm considering using flightplan for deployment (for a Laravel PHP project) however I'm not too sure if it's a good fit when it comes to the ability to easily do "rolling releases" and possibly rollback.

Take for instance a project called [deployer](https://github.com/deployphp/deployer) this creates a structure like so on the deployment/staging server:

```
/your/project/path
|--releases
|  |--20150513120631
|--shared
|  |--...
|--current -> /your/project/path/releases/20150513120631
```

Where `current` is the symlinked current release and `shared` is any shared resources between releases (image uploads, general documents, etc). [shipit](https://github.com/shipitjs/shipit) is another tool that similary does releases and symlinks etc.

I would like to achieve something similar in flightplan as I do like the API and simplicity of it, however I'm guessing I would have to create some kind of plugin to achieve the above? Are there any plans on adding some kind of "releases" type API? I think it makes a lot of sense in a deployment tool.

**Comment by pstadler** [comment]
One of the project goals behind flightplan is to keep it as simple as possible.

I agree that rolling releases is a feature in high demand and could probably be provided as some kind of official extension.

:+1:

**Comment by garygreen** [comment]
Yeah, I mean imo for it to be a proper deployment tool at minimum it needs to do some kind of provisioning on the server for releases. At the moment I see it as just a wrapper around SSH.

**Comment by pstadler** [comment]
That's the idea after all.



On Sat, Dec 12, 2015 at 2:35 PM -0800, "Gary Green" <notifications@github.com> wrote:












Yeah, I mean imo for it to be a proper deployment tool at minimum it needs to do some kind of provisioning on the server for releases. At the moment I see it as just a wrapper around SSH.

**Comment by JontyC** [comment]
There's another package which uses `flightplan` to do this sort of thing (albeit for AWS):

https://www.npmjs.com/package/cloudpirate

**Comment by evilebottnawi** [comment]
@pstadler maybe for such purposes we should think about interface to add plugins(extensions)? Add hook on basic action and(or) add config (also options for run their) for load extension(installed from npm or created local) containing complex tasks (example rolling).

**Comment by pstadler** [comment]
Absolutely.

**Comment by evilebottnawi** [comment]
@pstadler How about to create a separate issue and initiate a discussion about the hook (first step)?

**Comment by nikolay** [comment]
@pstadler Even if it still is going to stay as a simple SSH wrapper, it would be nice to have batching for things like deployments. For example, run a task N servers at a time and also have some rollback mechanism if things fail or a general stage-basd cleanup scripts. For example, regardless if you reach success or failure, if you create a temporary directory, it needs to be deleted at the end (except in some "intact" mode).

**Comment by evilebottnawi** [comment]
@nikolay good idea

**Comment by pstadler** [comment]
This is all possible with the current state of Flightplan. Remote tasks run in parallel and if something isn't working out you can roll back (e.g. having a try-catch in a remote flight which triggers some clean-up).

Such functionality should be provided as a plugin, because one of Flightplan's goals is to make as few assumptions as possible about how people deploy applications or operate their servers.

**Comment by garygreen** [comment]
> Such functionality should be provided as a plugin, because part of the philosophy behind Flightplan is to make as few assumptions as possible about how people deploy applications or operate their servers.

I would agree but the description of the project behind Flightplan states "deployment or systems administration tasks" -- It's absolutely reasonable to include in the core package of Flightplan some form of deploymenty stuff. It's fine to be somewhat opinionated about how that's done (as long as it's reasonable configurable) but it would make the project easier to use for actual deployment stuff straight out the box.

I was very confused (and somewhat frustrated) when I started using the project and couldn't easily use it for what it stated "deployments". Just my 2 cents :smile:

**Comment by nikolay** [comment]
@pstadler I personally don't mind if deployment is optional and when I said "rollbacks", I mean global, across the entire cluster, and not just individual rollback. For example, you may want to do an "all or nothing" type of update. If one server fails, all others may have to rollback.

**Comment by nikolay** [comment]
@pstadler Just to share a negative experience with the current state of Flightplan, which caused a brief Production outage for us. We needed to do a small Nginx config update and bounced all servers and because I didn't know this was always happening in parallel (my bad, of course). Essentially we ended up having no healthy web servers with the Load Balancer and our site went offline.
